libs/types-common/src/Data/Json/Util.hs

```diff
@@ -42,18 +48,37 @@ infixr 5 #
 
 -- | A newtype wrapper for 'UTCTime' that formats timestamps in JSON with
 -- millisecond precision instead of the default picosecond precision.
+-- Construct values using 'toUTCTimeMillis'; deconstruct with 'fromUTCTimeMillis'.
+-- Unlike with 'UTCTime', 'Show' renders ISO string.
 newtype UTCTimeMillis = UTCTimeMillis { fromUTCTimeMillis :: UTCTime }
-  deriving (Eq)
 
-instance Show UTCTimeMillis where
-    showsPrec d = showParen (d > 10) . showString . showUTCTimeMillis
+{-# INLINE toUTCTimeMillis #-}
+toUTCTimeMillis :: HasCallStack => UTCTime -> UTCTimeMillis
+toUTCTimeMillis = UTCTimeMillis . (TL.seconds %~ MkFixed . roundit . coerce @Pico @Integer)
+  where
+    roundit = (* onebillion) . (`div` onebillion)
```

Comment:
Fun fact: there is an extension (`-XNumDecimals`) that lets stuff like `1e9` have integral types.